search-sentiments-map/routes/search.js

```diff
@@ -0,0 +1,278 @@
+ * obtained every 12 hours for the specified topic.
+ */
+router.get('/:topic', (req, res) => {
+  let topic = req.params.topic;
+  retrieveSearchResultFromDatastore(topic).then(customSearchTopicJsonArray => {
+    // TODO(ntarn): Remove console.log statements when finished debugging.
+    console.log('ntarn debug: In the router.get method'); 
+    console.log(customSearchTopicJsonArray);
+    res.setHeader('Content-Type', 'application/json');
+    res.send(customSearchTopicJsonArray);
+  });
+});
+
+
+/** 
+ * Returns a JSON-formatted array of search results for countries retrieved
+ * from the Datastore.
+ * @param {string} topic Search topic to get data for.
+ */
+async function retrieveSearchResultFromDatastore(topic) {
+  // Request latest entity with a topic matching the given topic.
+  const query = datastore.createQuery('CustomSearchTopic').order('timestamp', {
+    descending: true,
+  }).filter('topic', topic).limit(1);
+
+  try {
+    const [customSearchTopic] = await datastore.runQuery(query);
+    let customSearchTopicJsonArray = {
+      topic: customSearchTopic[0].topic,
+      countries: customSearchTopic[0].countries,
+      timestamp: customSearchTopic[0].timestamp,
+    };
+    // TODO(ntarn): Remove console.log statements when finished debugging.
+    console.log(`ntarn debug retrieved search result for topic: ${topic}` +  `example: ${customSearchTopic[0].countries[0].results[0]}`);
+    console.log(customSearchTopicJsonArray);
+    return customSearchTopicJsonArray;
+  } catch (err) {
+    console.error('ERROR:', err);
+  }
+}
+
+/** 
+ * Updates daily search results (accumulates by day) in the Datastore.
+ */
+function updateSearchResults() {
+  retrieveGlobalTrends().then(async trends => {
+    for (let i = 0; i < trends.length; i++) { // When testing, make i < 1 trend to test one.
+      await updateSearchResultsForTopic(trends[i].trendTopic);
+      // TODO(ntarn): Remove console.log statements when finished debugging.
+      console.log('ntarn debug updateSearchResult for: ' + trends[i].trendTopic);
+      await new Promise(resolve => setTimeout(resolve, 60000));
+    }
+  });
+}
+
+/** 
+ * Retrieves search result data for all countries for a given topic. Passes
+ * this data to be saved in Datastore.
+ * @param {string} query Search query.
+ */
+async function updateSearchResultsForTopic(query) {
+  let countriesData = [];
+
+  // Note: Can't use forEach with await.
+  for (let i = 0; i < json.length; i++) { // When testing, make i < 3 countries. 
+    var set = 0;
+    // 100 queries per minute limit for Custom Search API. Pause to prevent
+    // surpassing limit.
+    if (i !== 0 && i % 100 === 0) {
+      await new Promise(resolve => setTimeout(resolve, 60000));
+    }
+    let countryData = [];
+    // Update countryData within the functions called.
+    const avg = await getSearchResultsForCountryFromAPI(
+      "country" + json[i].id, query, countryData);
```

Comment:
Honestly, I think the code would be cleaner if we return `countryData` from `getSearchResultsForCountryFromAPI` instead of passing it to layers of functions. (I was surprised that it works.)

It also seems strange to return avg (Btw rename this variable). I would only get the search results (countryData) with `getSearchResultsForCountryFromAPI` and get the sentiment by looping over the results obtained there.

Reply:
@carmenBen @ntarn This is probly the only technical comment here and we can discuss.

Reply:
Our goal was to only have to loop through the data once, rather than having to loop through it all again for the sentiment separately.

Reply:
I don't think looping through it again is a problem, and getting sentiments separately is cleaner. If you have to do it this way, at least rename the function here so that returning avg is less strange. @lawrencezjh @hayleejane3 Any thoughts?

Reply:
I needed it to return average in order for me to get the individual scores from addsearchresults in order to make the average, and to propagate those values through the methods that Carmen wrote. I also agree with Carmen, that we should only loop through the data once, it already takes an hour for the sentiment + search data to be updated itself. @carmenBen I'll let you rename the function as you see fit, I renamed the variable.

Reply:
I'm not sure why exactly we are renaming the function here.